### ISSUE TITLE
Problems with Apache ProxyPassReverse

### ISSUE THREAD

**Issue by KyleRoberts2a** [gollum/gollum#1107]
Hello,

I was looking into using gollum and could get it running just fine by itself.  Although trying to productionize it. I needed to put apache in front of it. 

I started by trying to put in a proxypass declaration like so

`ProxyPass /wiki http://localhost:4567/wiki
  ProxyPassReverse /wiki http://localhost:4567/wiki`

I did this in the main DocumentRoot which was /var/www. 

I created a bare repo which was owned by the www-data user and even fed gollum the wiki base path.

I tried many different options for a very long time. But the furthers I could get was apache complaining that

> create/wiki 

didn't exist.  I found this very odd since I am proxying /wiki and I can't say I have ever seen apache operate that way.

Any help would be appreciated.  Thanks so much. :)

**Comment by dometto** [comment]
I think you're looking for the `--base-dir` option -- see the readme. It would be great to have some example webserver configs in the wiki. Could you paste some examples in the wiki if you get it to work?

**Comment by dometto** [comment]
Sorry, that should be `--base-path`.

**Comment by KyleRoberts2a** [comment]
Thank you so much Dawa. I actually am setting the the following.

--base-path  /wiki

In which I am still getting the same errors. Should I need to do anything
else?

**Comment by SkyCrawl** [comment]
> I tried many different options for a very long time. But the furthers I could get was apache complaining that `create/wiki` didn't exist. I found this very odd since I am proxying /wiki and I can't say I have ever seen apache operate that way.

Indeed. I think we might be able to observe something from the logs, most likely the error and access logs. Could you post relevant parts of those in here after trying to access the Home page of your Gollum repository via your browser?

**Comment by KyleRoberts2a** [comment]
Logs were something I was looking at yesterday actually. It wasn't that the
create was coming before the base-path. But instead the following:



The collum webrick logs look like so

gollum --host 127.0.0.1 --port 4567 --base-path gollum
[2016-01-31 15:27:11] INFO  WEBrick 1.3.1
[2016-01-31 15:27:11] INFO  ruby 1.9.3 (2013-11-22) [x86_64-linux]
[2016-01-31 15:27:11] INFO  WEBrick::HTTPServer#start: pid=1231 port=4567
localhost - - [31/Jan/2016:15:27:17 EST] "GET /gollum HTTP/1.1" 404 18
- -> /gollum


Now I have tried many ways of proxying from apache.  Mostly like so

ProxyPass /gollum http://localhost:4567/gollum
ProxyPassReverse /gollum http://localhost:4567/gollum

The actual apache process sends it thorough and gets a 404. Access logs
look like so


ip - - [31/Jan/2016:15:51:09 -0500] "GET /gollum HTTP/1.1" 404 2328 "-"
"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101
Firefox/44.0"
ip - - [31/Jan/2016:15:51:34 -0500] "GET /gollum/ HTTP/1.1" 404 444 "-"
"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101
Firefox/44.0"
ip - - [31/Jan/2016:15:53:38 -0500] "GET /gollum HTTP/1.1" 404 1951 "-"
"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101
Firefox/44.0"

The only error logs are as follows


[Sun Jan 31 15:53:26.005084 2016] [ssl:warn] [pid 1735] AH01906: RSA server
certificate is a CA certificate (BasicConstraints: CA == TRUE !?)


But I believe that's due to my self signed cert. I get the same access 404s
if I don't use ssl redirection.

Here's a Gist to my apache2 configs.
https://gist.github.com/anonymous/ac981022ecd3fc8597b2


Thanks

**Comment by SkyCrawl** [comment]
> gollum --host 127.0.0.1 --port 4567 --base-path gollum
> localhost - - [31/Jan/2016:15:27:17 EST] "GET /gollum HTTP/1.1" 404 18 - -> /gollum

Looks like there goes our problem - that should NOT happen. I just tested it and got redirected to the "Create" page successfully (not tested with a reverse proxy though). So either the proxy messes up the request somehow (not likely) or there's something wrong with Gollum or your repository. What baffles me is that:
> I created a bare repo
> gollum --host 127.0.0.1 --port 4567 --base-path gollum

Your repository is probably bare but with the above launch command, Gollum is not supposed to support bare repositories when the default grit adapter is used (`--bare` flag should be needed for that). I tested the above launch command on a bare repository and as expected, I got a `Gollum::InvalidGitRepositoryError`.

In case your repository is not bare after all, we might be able to devise a way for Gollum to output some more useful info. To continue...

**Comment by dometto** [comment]
Can you confirm that it also doesnt work without the proxy, for example by connecting to localhost on the server machine (using `lynx` or `links`)?

**Comment by KyleRoberts2a** [comment]
well I can get to it if I change the listing address to the server IP
instead localhost.  I'll go and try localhost in a bit.

I have tried both bare and regular repos with the needed options and well
it didn't work :(

**Comment by SkyCrawl** [comment]
> In case your repository is not bare after all, we might be able to devise a way for Gollum to output some more useful info. To continue...

In case it's still relevant, let's try the following then:
* Add a `config.ru` file to your gollum repository with the contents below.
```ruby
#!/usr/bin/env ruby
require 'rubygems'
require 'sinatra'
require 'json'
require 'gollum/app'

# additional logging
class AppWithLoggedRequests < Precious::App
    use Rack::Logger
    # env['rack.errors']
    
    before do
        request.logger.info(JSON.pretty_generate(request.env.reject { |key, value| key.start_with?("rack")}))
    end
end

# base path handling, taken from the gollum binary
class MapGollum
    def initialize(base_path)
        @mg = Rack::Builder.new do
            map "/#{base_path}" do
                run AppWithLoggedRequests
            end
            map '/' do
                run Proc.new { [302, { 'Location' => "/#{base_path}" }, []] }
            end
            map '/*' do
                run Proc.new { [302, { 'Location' => "/#{base_path}" }, []] }
            end
        end
    end

    def call(env)
        @mg.call(env)
    end
end

# TODO: fill in the repo path:
gollum_path = "..."

# run
wiki_options = { :base_path => "gollum" }
AppWithLoggedRequests.set(:gollum_path, gollum_path)
AppWithLoggedRequests.set(:wiki_options, wiki_options)
Rack::Server.new(:app => MapGollum.new(wiki_options[:base_path]), :Port => 4567, :Host => '127.0.0.1').start
```
* Fill in path to your gollum repo in the above script at the designated place, and commit the file.
* Start gollum with the `rackup` command (working directory is the gollum repo).
* Start your reverse proxy.
* Try to access `http://localhost:80/gollum` from your browser.

Gollum will now dump request information to the console. With this, we should be able to see the exact difference between a working request (example below) and your 404 error. Ideally, Gollum should now output something like:
```
I, [2016-02-09T17:02:35.787094 #70501]  INFO -- : {
  "GATEWAY_INTERFACE": "CGI/1.1",
  "PATH_INFO": "",
  "QUERY_STRING": "",
  "REMOTE_ADDR": "127.0.0.1",
  "REMOTE_HOST": "localhost",
  "REQUEST_METHOD": "GET",
  "REQUEST_URI": "http://localhost:4567/gollum",
  "SCRIPT_NAME": "/gollum",
  "SERVER_NAME": "localhost",
  "SERVER_PORT": "4567",
  "SERVER_PROTOCOL": "HTTP/1.1",
  "SERVER_SOFTWARE": "WEBrick/1.3.1 (Ruby/2.1.8/2015-12-16)",
  "HTTP_HOST": "localhost:4567",
  "HTTP_USER_AGENT": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:44.0) Gecko/20100101 Firefox/44.0",
  "HTTP_ACCEPT": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
  "HTTP_ACCEPT_LANGUAGE": "cs,en-US;q=0.7,en;q=0.3",
  "HTTP_ACCEPT_ENCODING": "gzip, deflate",
  "HTTP_DNT": "1",
  "HTTP_CONNECTION": "keep-alive",
  "HTTP_VERSION": "HTTP/1.1",
  "REQUEST_PATH": "/gollum"
}
localhost - - [09/Feb/2016:17:02:35 CET] "GET /gollum HTTP/1.1" 302 0
- -> /gollum
I, [2016-02-09T17:02:35.810706 #70501]  INFO -- : {
  "GATEWAY_INTERFACE": "CGI/1.1",
  "PATH_INFO": "/Home",
  "QUERY_STRING": "",
  "REMOTE_ADDR": "127.0.0.1",
  "REMOTE_HOST": "localhost",
  "REQUEST_METHOD": "GET",
  "REQUEST_URI": "http://localhost:4567/gollum/Home",
  "SCRIPT_NAME": "/gollum",
  "SERVER_NAME": "localhost",
  "SERVER_PORT": "4567",
  "SERVER_PROTOCOL": "HTTP/1.1",
  "SERVER_SOFTWARE": "WEBrick/1.3.1 (Ruby/2.1.8/2015-12-16)",
  "HTTP_HOST": "localhost:4567",
  "HTTP_USER_AGENT": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:44.0) Gecko/20100101 Firefox/44.0",
  "HTTP_ACCEPT": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
  "HTTP_ACCEPT_LANGUAGE": "cs,en-US;q=0.7,en;q=0.3",
  "HTTP_ACCEPT_ENCODING": "gzip, deflate",
  "HTTP_DNT": "1",
  "HTTP_CONNECTION": "keep-alive",
  "HTTP_VERSION": "HTTP/1.1",
  "REQUEST_PATH": "/gollum/Home"
}
localhost - - [09/Feb/2016:17:02:35 CET] "GET /gollum/Home HTTP/1.1" 302 0
- -> /gollum/Home
I, [2016-02-09T17:02:36.876740 #70501]  INFO -- : {
  "GATEWAY_INTERFACE": "CGI/1.1",
  "PATH_INFO": "/create/Home",
  "QUERY_STRING": "",
  "REMOTE_ADDR": "127.0.0.1",
  "REMOTE_HOST": "localhost",
  "REQUEST_METHOD": "GET",
  "REQUEST_URI": "http://localhost:4567/gollum/create/Home",
  "SCRIPT_NAME": "/gollum",
  "SERVER_NAME": "localhost",
  "SERVER_PORT": "4567",
  "SERVER_PROTOCOL": "HTTP/1.1",
  "SERVER_SOFTWARE": "WEBrick/1.3.1 (Ruby/2.1.8/2015-12-16)",
  "HTTP_HOST": "localhost:4567",
  "HTTP_USER_AGENT": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:44.0) Gecko/20100101 Firefox/44.0",
  "HTTP_ACCEPT": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
  "HTTP_ACCEPT_LANGUAGE": "cs,en-US;q=0.7,en;q=0.3",
  "HTTP_ACCEPT_ENCODING": "gzip, deflate",
  "HTTP_DNT": "1",
  "HTTP_CONNECTION": "keep-alive",
  "HTTP_VERSION": "HTTP/1.1",
  "REQUEST_PATH": "/gollum/create/Home"
}
localhost - - [09/Feb/2016:17:02:36 CET] "GET /gollum/create/Home HTTP/1.1" 200 6705
- -> /gollum/create/Home
localhost - - [09/Feb/2016:17:02:38 CET] "GET /gollum/javascript/editor/langs/markdown.js?_=1455033758120 HTTP/1.1" 200 11837
http://localhost:4567/gollum/create/Home -> /gollum/javascript/editor/langs/markdown.js?_=1455033758120
```

**Comment by bartkamphorst** [comment]
@KyleRoberts2a Did you ever get this resolved?

**Comment by dmjacobsen** [comment]
I had a similar issue using Apache to perform reverse proxying to gollum, getting a 404 from any proxied URL.  This did not happen if I directly connected to the gollum instance, nor if I used nginx as the reverse proxy.  In the end I worked out that webrick was the core issue, by switching to another rack compatible server (thin, puma, etc) things started working.  Need to follow the Rack/Gollum guide to get that working: https://github.com/gollum/gollum/wiki/Gollum-via-Rack
